yktracec/src/jitmodbuilder.cc

```diff
@@ -1761,6 +1761,61 @@ class JITModBuilder {
           continue;
         }
 
+        if (Idx > 0 && Idx < InpTrace.Length() - 2) {
+          // Stores and loads into/from YkCtrlPointVars only need to be copied
+          // if they appear at the beginning or end of the trace. Any
+          // YkCtrlPointVars loads and stores inbetween come from tracing over
+          // the control point and aren't needed. We remove them here to reduce
+          // the size of traces.
+          if (isa<GetElementPtrInst>(I)) {
+            GetElementPtrInst *GEP = cast<GetElementPtrInst>(I);
+            if (GEP->getPointerOperand() == TraceInputs) {
+              // Collect stores
+              std::vector<Value *> Stores;
+              while (isa<GetElementPtrInst>(I)) {
+                assert(cast<GetElementPtrInst>(I)->getPointerOperand() == TraceInputs);
+                std::advance(I, 1);
+                assert(isa<StoreInst>(I));
+                Stores.insert(Stores.begin(), cast<StoreInst>(I)->getValueOperand());
+                std::advance(I, 1);
+              }
+              // Skip frameaddress, control point, and stackmap call
+              // instructions.
+              assert(isa<CallInst>(I)); // frameaddr call
+              std::advance(I, 1);
+              assert(isa<CallInst>(I)); // control point call
+              Value *CI = &*I;
+              std::advance(I, 1);
+              assert(isa<CallInst>(I)); // stackmap call
+              MPF->LastSMCall = cast<CallInst>(I);
+              std::advance(I, 1);
+              assert(isa<GetElementPtrInst>(I));
+              // Now find the loads and map them to the values of the previous
+              // stores.
+              while (isa<GetElementPtrInst>(I)) {
+                std::advance(I, 1);
+                assert(isa<LoadInst>(I));
+
+                VMap[&*I] = getMappedValue(Stores.back());
+                Stores.pop_back();
+                std::advance(I, 1);
+              }
+
+              // We've seen the control point so do the same we used to do when
+              // encountering a control point.
+              VMap[CI] = ConstantPointerNull::get(
+                  Type::getInt8PtrTy(JITMod->getContext()));
+
+              // The next two trace blocks are the (unmappable) control point
+              // followed by the return to this block which only has a branch
+              // instruction left which we don't care about. So we can just skip
+              // over these and continue with the block after.
+              Idx += 2;
```

Comment:
Can we add `assert`s just to check that the two instructions we skip are (at least) roughly of the form we expect?

Reply:
This instruction skips blocks, not instructions. Maybe @vext01 knows if there's something we can assert here.

Reply:
You can assert that they are unmapped, as long as you don't mind taking on the extra computation to do the mapping. Probably ok for debug builds.

Reply:
(I have a vague memory that mapping is currently eager anyway)

Reply:
Only one of the blocks is unmapped. The other is mapped (it's the block after returning from the control point).

Reply:
> Probably ok for debug builds.

Sounds OK to me!

Reply:
Perhaps the comment is wrong(?), but anyway, you can check that the mapability of the blocks is as expected, right?

Reply:
[724e819](https://github.com/ykjit/yk/pull/703/commits/724e819718e73e4786be3b01e55ad785a4aed4d9)